src/transformers/models/colpali/modular_colpali.py

```diff
@@ -0,0 +1,771 @@
+        num_logits_to_keep (`int`, *optional*):
+            Calculate logits for the last `num_logits_to_keep` tokens. If `0`, calculate logits for all
+            `input_ids` (special case). Only last token logits are needed for generation, and calculating them only for that
+            token can save memory, which becomes pretty significant for long sequences or large vocabulary size.
+
+    ```python
+    import torch
+    from PIL import Image
+
+    from transformers import ColPaliForRetrieval, ColPaliProcessor
+
+    model_name = "vidore/colpali-v1.2-hf"
+
+    model = ColPaliForRetrieval.from_pretrained(
+        model_name,
+        torch_dtype=torch.bfloat16,
+        device_map="cuda:0",  # or "mps" if on Apple Silicon
+    ).eval()
+
+    processor = ColPaliProcessor.from_pretrained(model_name)
+
+    # Your inputs
+    images = [
+        Image.new("RGB", (32, 32), color="white"),
+        Image.new("RGB", (16, 16), color="black"),
+    ]
+    queries = [
+        "What is the organizational structure for our R&D department?",
+        "Can you provide a breakdown of last year’s financial performance?",
+    ]
+
+    # Process the inputs
+    batch_images = processor(images=images).to(model.device)
+    batch_queries = processor(text=queries).to(model.device)
+
+    # Forward pass
+    with torch.no_grad():
+        image_embeddings = model(**batch_images)
+        query_embeddings = model(**batch_queries)
+
+    scores = processor.score_retrieval(query_embeddings, image_embeddings)
+    ```
+"""
+
+
+@add_start_docstrings(
+    """
+    ColPali leverages Vision Language Models (VLMs) to construct efficient multi-vector embeddings in the visual space for document retrieval.
+    By feeding the ViT output patches from PaliGemma-3B to a linear projection, we create a multi-vector representation of documents. The model
+    is trained to maximize the similarity between these document embeddings and the query embeddings, following the ColBERT method.
+
+    Using ColPali removes the need for potentially complex and brittle layout recognition and OCR pipelines with a single model that can take into account
+    both the textual and visual content (layout, charts, ...) of a document.
+
+    ColPali was introduced in the following paper: [*ColPali: Efficient Document Retrieval with Vision Language Models*](https://arxiv.org/abs/2407.01449).
+
+    Resources:
+    - A blog post detailing ColPali, a vision retrieval model, can be found [here](https://huggingface.co/blog/manu/colpali). 📝
+    - The code for using and training the original ColPali model and for the `colpali-engine` package can be found [here](https://github.com/illuin-tech/colpali). 🌎
+    - Cookbooks for learning to use the Hf version of ColPali, fine-tuning, and similarity maps generation can be found [here](https://github.com/tonywu71/colpali-cookbooks). 📚
+    """
+)
+class ColPaliForRetrieval(PaliGemmaForConditionalGeneration):
```

Comment:
I am not sure if this is the best approach. 
IMO we should just have a `self.model = AutoModel.from_config(PaliGemmaConfig())` + a linear layer on top ( as here it's not a new model, more like a task specific class no?

Reply:
I think your proposition makes sense as this would greatly simplify ColPali's `forward` method. I also think we could remove inheritance for `ColPaliConfig` and have a similar `config.vlm_backbone_config` attribute.

I believe it still makes sense to use `modular 🤗` for `class ColPaliProcessor(PaliGemmaProcessor)` because it feels anti-pattern to write `processor.processor`.

Wdyt?